# Machine Learning Zoomcamp - Homework #8

## Deep Learning 
#### Introduction to Neural Networks, Tensorflow and Keras

- In this homework I used a modified dataset containng images of bees and wasps. So, the main objective was to predict if a image contain a bee or a wasp using convolutional neural networks (CNN).


You can download the dataset for this homework from [here](https://github.com/SVizor42/ML_Zoomcamp/releases/download/bee-wasp-data/data.zip):

```bash
wget https://github.com/SVizor42/ML_Zoomcamp/releases/download/bee-wasp-data/data.zip
unzip data.zip
```

In [2]:
import tensorflow as tf
from tensorflow import keras
import scipy 

2023-11-16 19:49:38.862599: I tensorflow/core/util/util.cc:169] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.


## Question 1
- Since we have a binary classification problem, what is the best loss function for us?

A: Binary Crossentropy

In [3]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, InputLayer
from tensorflow.keras.optimizers import SGD

# creating a Sequential Model
model = Sequential()

# adding the input layer to the model
model.add(InputLayer(input_shape=(150, 150, 3)))

# Adding a convolutional layer with 32 filters, kernel size (3, 3) and activation 'relu'
model.add(Conv2D(32, (3, 3), activation='relu'))

# adding a max pooling layer with pooling size (2, 2)
model.add(MaxPooling2D((2, 2)))

# adding a flatten layer to convert the output to vectors
model.add(Flatten())

# adding a dense layer with 64 neurons and activation 'relu'
model.add(Dense(64, activation='relu'))

# adding a dense exit layer with 1 neuron and activation 'sigmoid'
model.add(Dense(1, activation='sigmoid'))

# Optimzer SGD with lr=0.002 and momentum=0.8
optimizer = SGD(learning_rate=0.002, momentum=0.8)

# compiling the model with binary crossentropy loss and accuracy metric
model.compile(optimizer=optimizer, loss='binary_crossentropy', metrics=['accuracy'])


2023-11-16 19:49:40.525499: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:975] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2023-11-16 19:49:40.532097: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:975] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2023-11-16 19:49:40.532320: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:975] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2023-11-16 19:49:40.532778: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F AVX512_VNNI FMA
To enable them in other operations, rebuild TensorFlow with the approp

In [4]:
# printing the model summary
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 148, 148, 32)      896       
                                                                 
 max_pooling2d (MaxPooling2D  (None, 74, 74, 32)       0         
 )                                                               
                                                                 
 flatten (Flatten)           (None, 175232)            0         
                                                                 
 dense (Dense)               (None, 64)                11214912  
                                                                 
 dense_1 (Dense)             (None, 1)                 65        
                                                                 
Total params: 11,215,873
Trainable params: 11,215,873
Non-trainable params: 0
____________________________________________

## Question 2
- What's the number of parameters in the convolutional layer of our model?

A: 11214912 

In [5]:
from tensorflow.keras.preprocessing.image import ImageDataGenerator
import scipy 
import numpy as np
# Create ImageDataGenerator for data augmentation and rescaling
datagen = ImageDataGenerator(rescale=1./255)

# Specify the paths to your training and test datasets
train_data_dir = 'train/'
test_data_dir = 'test/'

# specify the classes
desired_classes = ['bee', 'wasp']

# Set batch size
batch_size = 20

# Create generators for training and test data
train_generator = datagen.flow_from_directory(
    train_data_dir,
    target_size=(150, 150),
    batch_size=batch_size,
    class_mode='binary',  # for binary classification
    classes=desired_classes,
    shuffle=True
)

test_generator = datagen.flow_from_directory(
    test_data_dir,
    target_size=(150, 150),
    batch_size=batch_size,
    class_mode='binary',  # for binary classification
    classes=desired_classes,
    shuffle=True
)

# Train the model for 10 epochs
history = model.fit(
    train_generator,
    epochs=10,
    validation_data=test_generator
)

Found 2827 images belonging to 2 classes.
Found 918 images belonging to 2 classes.
Epoch 1/10


2023-11-16 19:49:42.627785: I tensorflow/stream_executor/cuda/cuda_dnn.cc:384] Loaded cuDNN version 8100
2023-11-16 19:49:43.235519: I tensorflow/core/platform/default/subprocess.cc:304] Start cannot spawn child process: No such file or directory
2023-11-16 19:49:43.236010: I tensorflow/core/platform/default/subprocess.cc:304] Start cannot spawn child process: No such file or directory
2023-11-16 19:49:43.236053: W tensorflow/stream_executor/gpu/asm_compiler.cc:80] Couldn't get ptxas version string: INTERNAL: Couldn't invoke ptxas --version
2023-11-16 19:49:43.236562: I tensorflow/core/platform/default/subprocess.cc:304] Start cannot spawn child process: No such file or directory
2023-11-16 19:49:43.236635: W tensorflow/stream_executor/gpu/redzone_allocator.cc:314] INTERNAL: Failed to launch ptxas
Relying on driver to perform ptx compilation. 
Modify $PATH to customize ptxas location.
This message will be only logged once.


142/142 [==============================] - 9s 48ms/step - loss: 0.6783 - accuracy: 0.5738 - val_loss: 0.6492 - val_accuracy: 0.6024
Epoch 2/10
142/142 [==============================] - 7s 47ms/step - loss: 0.6305 - accuracy: 0.6505 - val_loss: 0.6124 - val_accuracy: 0.6383
Epoch 3/10
142/142 [==============================] - 7s 47ms/step - loss: 0.5941 - accuracy: 0.6887 - val_loss: 0.5750 - val_accuracy: 0.6950
Epoch 4/10
142/142 [==============================] - 7s 47ms/step - loss: 0.5716 - accuracy: 0.7043 - val_loss: 0.5638 - val_accuracy: 0.7048
Epoch 5/10
142/142 [==============================] - 7s 48ms/step - loss: 0.5238 - accuracy: 0.7496 - val_loss: 0.5416 - val_accuracy: 0.7004
Epoch 6/10
142/142 [==============================] - 7s 47ms/step - loss: 0.4957 - accuracy: 0.7729 - val_loss: 0.5237 - val_accuracy: 0.7505
Epoch 7/10
142/142 [==============================] - 7s 47ms/step - loss: 0.4690 - accuracy: 0.7863 - val_loss: 0.5312 - val_accuracy: 0.7375
Epoch 8/10

In [6]:
# confirm the class indices
train_generator.class_indices

{'bee': 0, 'wasp': 1}

## Questions 3 and 4
- What is the median of training accuracy for all the epochs for this model?
- What is the standard deviation of training loss for all the epochs for this model?

In [7]:
# assign the train loss and accuracy to variables adn print them

train_loss = history.history['loss']
accuracy = history.history['accuracy']

print(f'The mean of accuracy was {round(np.mean(accuracy),2)} and the standard deviation of loss was {round(np.std(train_loss),3)}')

The mean of accuracy was 0.74 and the standard deviation of loss was 0.092


In [8]:
# data augmentation on training data

# Augmentations
train_datagen_aug = ImageDataGenerator(
    rescale=1./255,
    rotation_range=50,
    width_shift_range=0.1,
    height_shift_range=0.1,
    zoom_range=0.1,
    horizontal_flip=True,
    fill_mode='nearest'
)

# Create ImageDataGenerator for training with augmentations
train_aug_generator = train_datagen_aug.flow_from_directory(
    train_data_dir,
    target_size=(150, 150),
    batch_size=20,
    class_mode='binary',  
    classes=desired_classes,
    shuffle=True
)

Found 2827 images belonging to 2 classes.


In [9]:
# training the model with augmented data
aug_model = model.fit(
    train_aug_generator,
    epochs=10,
    validation_data=test_generator
)

Epoch 1/10
142/142 [==============================] - 18s 127ms/step - loss: 0.4967 - accuracy: 0.7634 - val_loss: 0.4756 - val_accuracy: 0.7756
Epoch 2/10
142/142 [==============================] - 18s 126ms/step - loss: 0.4886 - accuracy: 0.7768 - val_loss: 0.5854 - val_accuracy: 0.7113
Epoch 3/10
142/142 [==============================] - 18s 126ms/step - loss: 0.4801 - accuracy: 0.7754 - val_loss: 0.4712 - val_accuracy: 0.7810
Epoch 4/10
142/142 [==============================] - 18s 128ms/step - loss: 0.4751 - accuracy: 0.7825 - val_loss: 0.4980 - val_accuracy: 0.7691
Epoch 5/10
142/142 [==============================] - 18s 127ms/step - loss: 0.4642 - accuracy: 0.7913 - val_loss: 0.4741 - val_accuracy: 0.7865
Epoch 6/10
142/142 [==============================] - 18s 127ms/step - loss: 0.4582 - accuracy: 0.7927 - val_loss: 0.4855 - val_accuracy: 0.7647
Epoch 7/10
142/142 [==============================] - 18s 126ms/step - loss: 0.4539 - accuracy: 0.7885 - val_loss: 0.4796 - val_ac

## Question 5

- What is the mean of test loss for all the epochs for the model trained with augmentations?

In [10]:
# assign the test loss to a variable and print it
val_loss = history.history['val_loss']

print(f'The mean of test set loss was {round(np.mean(val_loss),3)}')

The mean of test set loss was 0.556


## Question 6

- What's the average of test accuracy for the last 5 epochs (from 6 to 10) for the model trained with augmentations?

In [11]:
# average accuracy for last 5 epochs on test set
last_five_accuracy = history.history['val_accuracy'][-5:]

print("Average Accuracy for last 5 epochs on test set:", round(np.average(last_five_accuracy), 3))

Average Accuracy for last 5 epochs on test set: 0.746
